### DISCUSSION TITLE
Why can't we enable oauth2 per consumer?

### DISCUSSION THREAD

**Discussion by nitin302** [Kong/kong#8820] [Help] [answered]
https://github.com/Kong/kong/blob/master/kong/plugins/oauth2/schema.lua#L18

Hello, as per the oauth2 schema, oauth2 can be enabled per service or route. However, I'd like to enable this per consumer, as other consumers may just require basic authentication or any other auth.

Why is this feature not given?

**Comment by Tieske** [comment, marked as answer]
That's not possible because the auth-plugins are required to identify the consumer. So when an auth plugin only runs on a consumer, how would it know when to run? Catch 22.

**Reply by nitin302** [reply]
normally yes. i'd agree but for oauth2 case (client_credentials flow), requires apikey to be given for generating a token and making the API call using the access_token.

Since, the apikey is added, consumer can be identified we can choose to/or not run oauth2 : https://support.konghq.com/support/s/article/How-to-use-the-OAuth-2-0-plugin-for-Client-Credential-flow

**Reply by nitin302** [reply]
Also, is there a way to enable auth-plugin by a tag instead of service/route. I am looking to enable oauth2 for certain consumers while others default to apikey.

**Reply by Tieske** [reply]
> but for oauth2 case (client_credentials flow), requires apikey to be given for generating a token and making the API call using the access_token.

Currently oauth2 cannot do this. But PR's would be considered.
